Graduate data view options out of a menu to allow more design expression

jorgefilipecosta opened the pull request:
This PR implements the new view config UI proposed at https://github.com/WordPress/gutenberg/issues/63872.
To keep this PR manageable and not very big we just port the existing functionality of the old dropdown menu. (Change some parameters, and hide/show fields).
Things like moving the item size picker for the grid into the popover, or sort directly on the field list will come as follow-ups.

This PR tries to implement the patterns specified in https://github.com/WordPress/gutenberg/issues/63624.
SettingsSection has container queries to make the component responsive depending on the size of the container.

## Screenshots
<img width="464" alt="Screenshot 2024-08-01 at 17 11 34" src="https://github.com/user-attachments/assets/2629a383-0956-4920-a839-48f2d07690dd">
<img width="842" alt="Screenshot 2024-08-01 at 17 12 04" src="https://github.com/user-attachments/assets/30d7ebe2-d404-4301-bfc9-62fa57ca795f">

cc: @jameskoster 


## Testing Instructions
I verified I could configure all the fields and things worked as expected.
Using the browser inspector I added the following CSS rule:
```
.dataviews-view-config {
    width: 650px !important;
}
```
I verified I got the responsive behavior shown above.

jorgefilipecosta committed:
New view config UI

jorgefilipecosta committed:
lint fix

youknowriad commented:
Great start. I've been testing this and it works well.

 - Personally, I'm not sure the "layout type" should be part of the dropdown though. It deserves more prominence and also, it will have an impact on the dropdown's content later. 
 - I found myself missing the reordering feature (arrows to move fields)
 - Do you expect to implement the grid density slider in the dropdown in a follow-up PR?

youknowriad commented on packages/dataviews/src/components/dataviews-view-config/style.scss:
```diff
@@ -0,0 +1,39 @@
+	font-size: $helptext-font-size;
+	font-style: normal;
+	color: $gray-700;
+}
```
I'm a bit surprised by the number of styles that we need to make this look correct. Do we need all these "widths", "margins", "paddings"... Why our default components don't work well? What's missing as a UI component to render that dropdown?

youknowriad replied:
Reading the code again, it seems mostly about the `SettingsSection` component. I would suggest extract this component to its own folder `dataviews/src/components/fieldset` or something to start with and IMO, this should be a UI component like a `Fieldset` component or something. I also wonder if it should be using a `fieldset + legend` element (not entirely sure)

tyxla replied:
We could be able to save some of these styles by using the `Text` component.

jameskoster commented:
Nice work Jorge, this is working really well already imo.

> Personally, I'm not sure the "layout type" should be part of the dropdown though. It deserves more prominence and also, it will have an impact on the dropdown's content later.

Yes, I don't think we should remove the top-level layout switcher. 

The thinking behind including it in the popover is that it would give context to some of the layout-specific options that appear conditionally, like grid density. However seeing it in practise the jump when switching to list layout is a bit extreme. 

Let's leave the layout switcher as it is for now.

> I found myself missing the reordering feature (arrows to move fields)

This isn't on trunk, so probably fine to add separately? If we do, we should update the description so it doesn't mention re-ordering :) 

@jorgefilipecosta instead of toggling the opacity on the visibility button, could we toggle the icon instead (between `seen` and `unseen`)?

The spacing in the `ItemGroup` is a bit clunky. It's not something to address here, but in the component. I know it's on @WordPress/gutenberg-components' list, but wanted to note another occurrence. 

---

I'd appreciate feedback from @WordPress/gutenberg-design on the idea of referring to fields as 'properties'. It feels more user-friendly to me, especially with one eye on content modelling.

jameskoster committed:
Style adjustments

jameskoster commented:
I pushed a couple of minor style adjustments.

jameskoster commented:
@jorgefilipecosta one small styling discrepancy I couldn't figure out... the grid seems to be correctly, but the dimensions are off. Notice that the inputs are narrower than the title + description section:

<img width="1155" alt="Screenshot 2024-08-02 at 18 03 45" src="https://github.com/user-attachments/assets/b6249702-24c3-46c3-a4e2-72c2d65cc12f">

jorgefilipecosta commented:
Hi @jameskoster, good catch.
The fields are not using a grid.
But title/description and fields use a grid with 4fr 8fr dimensions. If the fields were correctly divided in two they would also have a 4fr dimension being the same size as title/description. We are probably missing some padding margin on a calculation somewhere I can look into this.

jameskoster commented:
Maybe we can use css sub grid? [Here's a codepen](https://codepen.io/jameskoster/pen/gONWpzP?editors=1100) that seems to work.

<img width="1081" alt="Screenshot 2024-08-02 at 20 49 40" src="https://github.com/user-attachments/assets/ddb704a5-eb68-4983-8000-441ed6984dab">

tyxla commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -156,186 +201,122 @@ function FieldsVisibilityMenu< Item >( {
+		<VStack className="dataviews-view-config" spacing={ 6 }>
+			<SettingsSection
+				title={ __( 'Appearance' ) }
+				description={ __( 'Customise the display of data' ) }
```
We're generally using US English:

```suggestion
				description={ __( 'Customize the display of data' ) }
```

tyxla commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -1,150 +1,195 @@
+							...defaultLayouts[ newLayout ],
+						} );
+				}
+				throw new Error( 'Invalid dataview' );
```
Do we really want to throw errors in an `onChange` handler? I'd expect a `warning()` or something less invasive. I see this was here before, so just something to keep in mind.

jorgefilipecosta replied:
Good point, I followed what we had without thinking much but a warning seems like a better option. The code was updated 👍

tyxla commented on packages/dataviews/src/constants.ts:
```diff
@@ -58,6 +58,10 @@ export const sortLabels = {
 	desc: __( 'Sort descending' ),
 };
+export const sortLabelsShort = {
+	asc: __( 'ASC' ),
```
AFAIK we rarely capitalize strings; instead, capitalization is achieved cosmetically with CSS.

Also, note that this can be problematic for translation in some languages. If we can, I'd recommend we use the full words (ascending / descending)

jorgefilipecosta replied:
I updated the code to use CSS as suggested. Regarding using full words (ascending / descending), in this case, it would break the design: 
<img width="447" alt="Screenshot 2024-08-06 at 14 06 21" src="https://github.com/user-attachments/assets/960384a3-a39c-429d-8b1d-969f92acf3c7">

I added a translation context, that also specifies a character limit for the short labels.

tyxla replied:
@jorgefilipecosta thanks for the screenshot. This is exactly what will happen with the translations, if folks start translating and there's no good abbreviation in their language.

I wonder if a worthy alternative is to use icons instead of ASC/DESC text labels. We have a few up/down arrow icons available that we could use here. @jameskoster WDYT?

jameskoster replied:
Yup, let's try an icon :)

tyxla commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -156,186 +201,122 @@ function FieldsVisibilityMenu< Item >( {
-		</DropdownMenu>
+		<Grid columns={ 2 } className="dataviews-settings-section" gap={ 4 }>
+			<VStack spacing={ 0 }>
+				<h2 className="dataviews-settings-section__title">{ title }</h2>
```
Maybe we can use the `__experimentalHeading` component for these headings? Could perhaps save some custom styles.

jorgefilipecosta replied:
The code was updated and indeed it allows us to remove some custom css.

tyxla reviewed (commented):
I've left some drive-by comments as I was responding to a specific ping.

> The spacing in the `ItemGroup` is a bit clunky. It's not something to address here, but in the component. I know it's on @WordPress/gutenberg-components' list, but wanted to note another occurrence.

@jameskoster could you please elaborate on the spacing issues you're seeing? Yes, we've been meaning to do some unification of the menu and item components (see #35210) plus some general flexibility and integration improvements to `ItemGroup` (see #34709) but I'm not sure we're aware of any spacing issues, specifically.

jameskoster commented:
> @jameskoster could you please elaborate on the spacing issues you're seeing?

@tyxla it relates mostly to the dimensions. Most interactive elements follow a 24/32/40px sizing scale for height, but due to the unconventional padding on `Item` we often see peculiar heights. In this case each `Item` is `48.8px`. In Global Styles, the `Item`s in shadow management and 'Elements' UIs are `41.8px` tall. `Item`s in font lists are `44px`. And so on. It's all just very inconsistent.

jorgefilipecosta committed:
refactor grid

jorgefilipecosta committed:
address reviews

jorgefilipecosta committed:
add traslation context

jorgefilipecosta commented:
> Maybe we can use css sub grid? [Here's a codepen](https://codepen.io/jameskoster/pen/gONWpzP?editors=1100) that seems to work.
> 
> <img alt="Screenshot 2024-08-02 at 20 49 40" width="1081" src="https://private-user-images.githubusercontent.com/846565/354708717-ddb704a5-eb68-4983-8000-441ed6984dab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NDg0NDIsIm5iZiI6MTcyMjk0ODE0MiwicGF0aCI6Ii84NDY1NjUvMzU0NzA4NzE3LWRkYjcwNGE1LWViNjgtNDk4My04MDAwLTQ0MWVkNjk4NGRhYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNlQxMjQyMjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTk0M2UxNjVmYjc5NjA2NGQ0NWM4ODBhMTBjZTUxNDQ5NjI1MThiZDIzNTUyZjM1ZDM3NDZlMjM5OTFlMDQzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.XeKjnUaaFPP1XESz4MOSd4AFf5vuh8VFw_XvU1S0FoM">

Thank you a lot for the grid sample provided @jameskoster. I updated the code to use something similar to what you suggested and the dimensions issue you noticed is now fixed.

jorgefilipecosta commented:
Thank you for the review @tyxla, I have addressed the feedback provided.

jorgefilipecosta committed:
lint fixes

jorgefilipecosta committed:
use unseen icon

jorgefilipecosta commented:
> Great start. I've been testing this and it works well.
> 
> * Personally, I'm not sure the "layout type" should be part of the dropdown though. It deserves more prominence and also, it will have an impact on the dropdown's content later.
> * I found myself missing the reordering feature (arrows to move fields)
> * Do you expect to implement the grid density slider in the dropdown in a follow-up PR?



Hi @youknowriad, the PR was updated and the layout type was left as is.

I will prioritize working on reordering and grid density in the popover as follow-up tasks. The main objective of this PR is to simply implement the popover with the existing functionality of the menu without making it too extensive.

jorgefilipecosta commented:
Hi @jameskoster,

> @jorgefilipecosta instead of toggling the opacity on the visibility button, could we toggle the icon instead (between seen and unseen)?

I made that update, I think this PR should be ready for another look.

youknowriad commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -8,143 +8,216 @@ import type { ChangeEvent } from 'react';
+			__nextHasNoMarginBottom
+			__next40pxDefaultSize
+			isBlock
+			label={ __( 'Order' ) }
```
If I click the "sort by" label, the select is focused properly, so the label is working properly but it's not the case for "Order" or "Items per page". I expect the button group (or the first button) to be focused. It seems we might have an issue in the component. cc @tyxla

mirka replied:
I have thought about this, but I'm not sure on what basis we would want to implement that behavior. Usually we do it for accessibility (i.e. increase the click target on things like checkboxes and radios, or because "the browser does it by default" when an `input` is properly associated with a `label`). In the case of compound components like ToggleGroupControl, I don't think there's a click target issue, and since in normal HTML semantics the label would actually be associated with a `radiogroup`/`group`, or be a `legend` in a `fieldset`, that kind of click-to-focus behavior would not happen.

So the only reason to add click-label-to-focus on compound components would be for UX reasons (consistency), which may possibly be bad for accessibility due to being non-standard 🤔

tyxla replied:
The behavior here is expected AFAIK, because the underlying`ToggleGroupControl` behaves like a radiogroup with radio buttons. You can see an example with actual radio buttons here: https://ariakit.org/components/radio where you can tab into and out of the component with `Tab` / `Shift` + `Tab`, and you can use the arrows to select a different radio button.

If you're referencing the focus styles, this is another thing: we've been debating this in https://github.com/WordPress/gutenberg/pull/63450#issuecomment-2228354818 and decided to not move the focus to the `ToggleGroupControlOption` because, it a nutshell, it would look odd with the selected style.  We can always re-evaluate that if you feel like it was not the best path forward.

youknowriad replied:
> So the only reason to add click-label-to-focus on compound components would be for UX reasons (consistency), which may possibly be bad for accessibility due to being non-standard 🤔

Yes, that's my point for a user, it's a single control, just like the others in the form. It's not really clear why there's a different between these controls and the "sort by" for instance. 

I can live with that for now, but for me it's still a bug and there should be a way to fix the UX without breaking the a11y IMO.

jameskoster commented:
One small change; we can probably leave out the descriptions for now. I tried to remove them but got an error. To be clear; in `SettingsSection` the description should be optional :)

jameskoster commented on packages/dataviews/src/components/dataviews-view-config/style.scss:
```diff
@@ -0,0 +1,43 @@
+.dataviews-view-config {
+	width: 363px;
```
Let's make this a bit narrower for now, and align to the 4px baseline. `320px` works well.

jameskoster commented on packages/dataviews/src/components/dataviews-view-config/style.scss:
```diff
@@ -0,0 +1,43 @@
+
+.dataviews-settings-section__title.dataviews-settings-section__title {
+	line-height: $grid-unit-30;
+    font-size: 15px;
```
Not something to address here, but it would be great to update the `Text` / `Heading` components to support type styles so that we can avoid this kind of css. cc @tyxla.

tyxla replied:
I believe both the `Text` and the `Heading` components accept a `size` prop that you can use:

```
<Heading size={ 18 } />My heading</Heading>
<Text size={ 13 } />My text</Text>
```

Now, I can argue whether doing this is a good idea. Specifying an arbitrary size can work against our efforts to bring consistency across the editor. Ideally, you would use one of the available `level` values of `Heading` (1 to 6), and `Text` would be consistent across the editor. 

cc @WordPress/gutenberg-components in case anyone else has further thoughts.

jameskoster replied:
Yup, ideally we move away from arbitrary sizes entirely. I know it's no small task though.

ciampo replied:
What @jameskoster said 💯

jorgefilipecosta committed:
enhacements

jorgefilipecosta committed:
lint fixes

jorgefilipecosta committed:
lint and end to end test updates

jorgefilipecosta committed:
update package lock

jorgefilipecosta committed:
test update

jorgefilipecosta committed:
test update

jorgefilipecosta committed:
lint fix

jorgefilipecosta committed:
end to end test update

jorgefilipecosta committed:
end to end test update

jorgefilipecosta committed:
lint fix

youknowriad commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -156,185 +231,124 @@ function FieldsVisibilityMenu< Item >( {
+					onClick={ () => setIsShowingViewPopover( true ) }
+				/>
+				{ isShowingViewPopover && (
+					<Popover
```
Do you know why we're using `Popover` (too low level) here, instead of `Dropdown` ?

The reason I'm asking is that I see a lot of differences between this dropdown and the "layout type dropdown"

 - The spacing is not the same between the anchor and the popover 
 - The behavior is not the same: one dropdown disabled everything behind it (inhert, disables scrolling...) and the other doesn't
